test/chainbase_test.cpp

```diff
@@ -0,0 +1,506 @@
+   template<typename Constructor, typename Allocator>
+   account(Constructor&& c, Allocator&& a) { c(*this); }
+   id_type id;
+   arbitrary_datum _account_key;
+   arbitrary_datum _account_value;
+};
+
+using account_index = multi_index_container<
+  account,
+  indexed_by<
+    ordered_unique<member<account,account::id_type,&account::id>>,
+    ordered_non_unique<BOOST_MULTI_INDEX_MEMBER(account,arbitrary_datum,_account_key)>,
+    ordered_non_unique<BOOST_MULTI_INDEX_MEMBER(account,arbitrary_datum,_account_value)>
+  >,
+  chainbase::allocator<account>
+>;
+
+CHAINBASE_SET_INDEX_TYPE(account, account_index)
+
+class database_test {
+public:
+   database_test(size_t lower_bound_inclusive,
+                 size_t upper_bound_inclusive,
+                 size_t num_of_accounts_and_values,
+                 size_t num_of_swaps,
+                 size_t max_key_length,
+                 size_t max_key_value,
+                 size_t max_value_length,
+                 size_t max_value_value)
+      : _gen_data{lower_bound_inclusive,
+                  upper_bound_inclusive,
+                  num_of_accounts_and_values,
+                  num_of_swaps,
+                  max_key_length,
+                  max_key_value,
+                  max_value_length,
+                  max_value_value}
+   {
+      _database.add_index<account_index>();
+   }
+
+   ~database_test()
+   {
+      boost::filesystem::remove_all(_database_dir);
+   }
+
+   inline void start_test() {
+      _initial_database_state();
+      _execution_loop();
+      loggerman.flush_all();
+   }
+
+private:
+   const boost::filesystem::path _database_dir{boost::filesystem::current_path() /= std::string{"/chainbase"}};
+   chainbase::database _database{_database_dir, chainbase::database::read_write, static_cast<uint64_t>((uint64_t)4096*(uint64_t)100000000)};
+   generated_data _gen_data;
+   system_metrics _system_metrics;
+
+   inline void _initial_database_state() {
+      size_t transactions_per_second{};
+
+      time_t old_time{initial_time};
+      time_t new_time{initial_time};
+      
+      std::cout << "Filling initial database state...\n" << std::flush;
+      loggerman.print_progress(1,0);
+      for (size_t i{}; i < _gen_data.num_of_accounts_and_values()/10; ++i) {
+         new_time = time(NULL);
+         if (new_time != old_time) {
+            loggerman.log_tps      ({(new_time - initial_time), transactions_per_second/(new_time - initial_time)});
+            loggerman.log_ram_usage({(new_time - initial_time), _system_metrics.total_ram_currently_used()});
+            loggerman.log_cpu_load ({(new_time - initial_time), _system_metrics.calculate_cpu_load()});
+            loggerman.print_progress(i, _gen_data.num_of_accounts_and_values()/10);
+            old_time = new_time;
+         }
+
+         _database.start_undo_session(true);
+         
+         auto session{_database.start_undo_session(true)};
+         _database.start_undo_session(true);
```

Comment:
Why start so many undo sessions?

Reply:
My reasoning is to simulate an already long chain of `undo_session`s (blocks) for the test to operate on. In this case a new block is pushed for ever 10 accounts/balances created.